server/src/main/java/org/elasticsearch/index/search/stats/FieldUsageStats.java

```diff
@@ -0,0 +1,279 @@
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class FieldUsageStats implements ToXContentFragment, Writeable {
+    public static final String TOTAL = "total";
+    public static final String TERMS = "terms";
+    public static final String FREQS = "frequencies";
+    public static final String POSITIONS = "positions";
+    public static final String OFFSETS = "offsets";
+    public static final String DOC_VALUES = "doc_values";
+    public static final String STORED_FIELDS = "stored_fields";
+    public static final String NORMS = "norms";
+    public static final String PAYLOADS = "payloads";
+    public static final String TERM_VECTORS = "term_vectors"; // possibly refine this one
+    public static final String POINTS = "points";
+    public static final String PROXIMITY = "proximity";
+
+    private final Map<String, PerFieldUsageStats> stats;
+
+    public FieldUsageStats() {
+        this.stats = new HashMap<>();
+    }
+
+    public FieldUsageStats(Map<String, PerFieldUsageStats> stats) {
+        this.stats = new HashMap<>(stats);
+    }
+
+    public FieldUsageStats(StreamInput in) throws IOException {
+        stats = in.readMap(StreamInput::readString, PerFieldUsageStats::new);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap(stats, StreamOutput::writeString, (o, v) -> v.writeTo(o));
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("all_fields");
+        total().toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject("fields");
+        {
+            final List<String> sortedFields = stats.keySet().stream().sorted().collect(Collectors.toList());
+            for (String field : sortedFields) {
+                builder.startObject(field);
+                stats.get(field).toXContent(builder, params);
+                builder.endObject();
+            }
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    PerFieldUsageStats total() {
+        final PerFieldUsageStats total = new PerFieldUsageStats();
+        for (PerFieldUsageStats value : stats.values()) {
+            total.add(value);
+        }
+        return total;
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    public boolean hasField(String field) {
+        return stats.containsKey(field);
+    }
+
+    public PerFieldUsageStats get(String field) {
+        return stats.get(field);
+    }
+
+    public FieldUsageStats add(FieldUsageStats other) {
+        other.stats.forEach((k, v) -> stats.computeIfAbsent(k, f -> new PerFieldUsageStats()).add(v));
+        return this;
+    }
+
+    public enum UsageContext {
+        DOC_VALUES,
+        STORED_FIELDS,
+        TERMS,
```

Comment:
should we have `postings` in addition to `terms`? E.g. the new `_terms_enum` API would consume terms but not postings

Reply:
Is this distinction relevant to a user? As far as I'm aware, they can't configure one without the other. I was also wondering if it's good enough to say that "postings" have been used whenever `TermsEnum.postings()` has been called (or if we can only truly know this once `nextDoc` or `docId` is called, which would be problematic to track efficiently)?